src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,160 @@
+    member private this.TryGetItemsToEvict () =
+        match options.Strategy with
+        | CachingStrategy.LRU ->
+            this.Store.Values |> Seq.sortBy _.LastAccessed |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+        | CachingStrategy.MRU ->
+            this.Store.Values |> Seq.sortByDescending _.LastAccessed |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+        | CachingStrategy.LFU ->
+            this.Store.Values |> Seq.sortBy _.AccessCount |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+
+    member private this.TryEvictItems () =
+        if this.GetEvictCount() > 0 then
+            for key in this.TryGetItemsToEvict () do
+                let (removed, _) = this.Store.TryRemove(key)
+                if removed then
+                    eviction.Trigger(key, EvictionReason.Evicted)
+
+    member private this.TryEvictTask () =
+        // This will spin in the background trying to evict items.
+        // One of the issues is that if the delay is high (>100ms), it will not be able to evict items in time, and the cache will grow beyond the maximum capacity.
+        backgroundTask {
+            while not cts.Token.IsCancellationRequested do
+                if this.GetEvictCount() > 0 then
+                    this.TryEvictItems ()
+                // do! Task.Delay(100, cts.Token)
+        }
+
+    // TODO: Explore an eviction shortcut, some sort of list of keys to evict first, based on the strategy.
+    member this.TryEvict() =
+        if this.GetEvictCount() > 0 then
+            match options.EvictionMethod with
+            | EvictionMethod.Blocking -> this.TryEvictItems ()
+            | EvictionMethod.Background -> ()
+
+    member _.TryGet(key: 'Key) =
+        match this.Store.TryGetValue(key) with
+        | true, value ->
+            // this is fine to be non-atomic, I guess, we are okay with race if the time is within the time of multiple concurrent calls.
+            value.LastAccessed <- DateTimeOffset.Now.Ticks
+            value.AccessCount <- Interlocked.Increment(&value.AccessCount)
+            cacheHit.Trigger(key, value.Value)
+            ValueSome value
+        | _ ->
+            cacheMiss.Trigger(key)
+            ValueNone
+
+    [<NoCompilerInlining>]
+    member this.Add(key: 'Key, value: 'Value) = let _ = this.TryAdd(key, value) in ()
+
+    [<NoCompilerInlining>]
+    member this.TryAdd<'Key>(key: 'Key, value: 'Value) =
+
+        // Weak table/references only make sense if we work with reference types (for obvious reasons).
+        // So, if `Weak` is enabled for value types as keys of the cache, it will simply box them, resulting in additional allocations.
+        // GC-based eviction shall not be used with value types as keys.
+        if options.Weak then
+            let weak = Weak<'Key>(key)
+            this.ConditionalWeakTable.TryAdd(key :> obj, weak) |> ignore
+            weak.Collected.Add(this.RemoveCollected)
```

Comment:
This will not be collected, because there is another reference to `key` from the inner storage of `this.Store`

Reply:
I can see them getting collected if I trigger the GC. I.e. I see finalizer getting called on them.

Reply:
Ah, disregard, I thought I've pushed updated version which wraps key in a weak reference. Will push it on Monday.